### DISCUSSION TITLE
TornadoVM v0.15.2

### DISCUSSION THREAD

**Discussion by jjfumero** [beehive-lab/TornadoVM#265] [announcements]
Date: 26/07/2023

## Improvements

- Initial Support for Multi-Tasks on Multiple Devices (MTMD): This mode enables the execution of multiple independent tasks on more than one hardware accelerator. Documentation in link: https://tornadovm.readthedocs.io/en/latest/multi-device.html
- Support for trigonometric ``radian``, ``cospi`` and ``sinpi`` functions for the OpenCL/PTX and SPIR-V backends.
- Clean-up Java modules not being used and TornadoVM core classes refactored.

## Compatibility/Integration

- Initial integration with ComputeAorta (part of the Codeplay's oneAPI Construction Kit for RISC-V) to run on RISC-V with Vector Instructions (OpenCL backend) in emulation mode.  
- Beehive SPIR-V Toolkit dependency updated.
- Tests for prebuilt SPIR-V kernels fixed to dispatch SPIR-V binaries through the Level Zero and OpenCL runtimes. 
- Deprecated ``javac.py`` script removed. 

## Bug fixes:

- TornadoVM OpenCL Runtime throws an exception when the detected hardware does not support FP64.
- Fix the installer for the older Apple with the x86 architecture using AMD GPUs. 
- Installer for ARM-based systems fixed.
- Installer fixed for Microsoft WSL and NVIDIA GPUs. 
- OpenCL code generator fixed to avoid using the reserved OpenCL keywords from Java function parameters.
- Dump profiler option fixed. 

<hr /><em>This discussion was created from the release <a href='https://github.com/beehive-lab/TornadoVM/releases/tag/v0.15.2'>TornadoVM v0.15.2</a>.</em>

**Comment by lucas8338** [comment]
instalation stcript doesnt work. i got error 404 when the script tries to download maven 3.9.3.
i'm trying to install tornadovm on WSL1. i think this error is happening because the maven has removed the "3.9.3" version
from their release repository. "https://dlcdn.apache.org/maven/maven-3".

```
lucas@WIN:~/tornado/TornadoVM-0.15.2/scripts$ ./tornadovm-installer --jdk jdk17 --backend opencl
Checking dependency: cmake-3.25.2-linux-x86_64.tar.gz

Checking dependency: apache-maven-3.9.3-bin.tar.gz
Traceback (most recent call last):
File "/home/lucas/tornado/TornadoVM-0.15.2/scripts/./tornadovm-installer", line 355, in
installer.install(args)
File "/home/lucas/tornado/TornadoVM-0.15.2/scripts/./tornadovm-installer", line 293, in install
self.downloadMaven()
File "/home/lucas/tornado/TornadoVM-0.15.2/scripts/./tornadovm-installer", line 144, in downloadMaven
wget.download(url, fullPath)
File "/home/lucas/.local/lib/python3.10/site-packages/wget.py", line 526, in download
(tmpfile, headers) = ulib.urlretrieve(binurl, tmpfile, callback)
File "/usr/lib/python3.10/urllib/request.py", line 241, in urlretrieve
with contextlib.closing(urlopen(url, data)) as fp:
File "/usr/lib/python3.10/urllib/request.py", line 216, in urlopen
return opener.open(url, data, timeout)
File "/usr/lib/python3.10/urllib/request.py", line 525, in open
response = meth(req, response)
File "/usr/lib/python3.10/urllib/request.py", line 634, in http_response
response = self.parent.error(
File "/usr/lib/python3.10/urllib/request.py", line 563, in error
return self._call_chain(*args)
File "/usr/lib/python3.10/urllib/request.py", line 496, in _call_chain
result = func(*args)
File "/usr/lib/python3.10/urllib/request.py", line 643, in http_error_default
raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found
lucas@WIN:~/tornado/TornadoVM-0.15.2/scripts$
```

**Reply by jjfumero** [reply]
Hi @lucas8338 , yes the links changed for the maven dependency. Some workaround:
1. You can install the dependencies and then compile TornadoVM. We have documentation for this  installation: https://tornadovm.readthedocs.io/en/latest/installation.html#b-manual-installation 
2. The second option is to use the latest version in the `master` branch. The API is the same and there are many fixes.

**Comment by lucas8338** [comment]
i fixed this problem changing the maven repo links in the /scripts/installerConfig.py

lines: 62, 63. (maybe only the 62 is needed)

to:
```
## Maven 
MAVEN = {
    __LINUX__ : {
        __X86_64__ : "https://dlcdn.apache.org/maven/maven-3/3.9.4/binaries/apache-maven-3.9.4-bin.tar.gz", #62
        __ARM__    : "https://dlcdn.apache.org/maven/maven-3/3.9.4/binaries/apache-maven-3.9.4-bin.tar.gz", #63
    },
    __APPLE__ : {
        __X86_64__ : None,   
        __ARM__    : None,
    }
}
```

also is important to know that to install tornadovm (thought the "tornadovm-installer" script) you need to be in the root directory of the tornadovm. running the "tornadovm-installer" while inside the "/scripts" directory return fake results (you get installation successful message but is fake).

i also discovered is not possible to use gpu in wsl1 (maybe is possible in wsl2 but my computer is too much old for wsl2) so i need to install a linux distro in my computer. (i have tried msys2 and cygwin but i got problems about python).

**Reply by jjfumero** [reply]
Thank you @lucas8338 for the report. We fixed the links already in the mater branch. Regarding the WSL, yes, you need WSL 2. I am using Windows 11 and WSL2 and it works for SPIR-V, OpenCL and PTX for GPUs and CPUs. 

What is the error you get with Python? Note that we use the Msys64 terminal.

**Reply by lucas8338** [reply]
the problem about python was conficts (i guess) among computer environment variables and Msys64 python instalation i believe Msys64 was trying to use an inexistent python causing a recursive problem about the location of python, but at this point i was trying to use Msys64 to run the "tornadovm-installer" script.

**Comment by lucas8338** [comment]
sucesful installation trought "https://tornadovm.readthedocs.io/en/latest/installation.html#installation-windows" on Msys64 (mingw64 env), win10, maven 3.9.4, "graalvm-jdk-20.0.2+9.1", gt710, opencl.

just make sure graalvm (or open jdk etc.) doenst can be located in any directory with empty spaces (example: C:/Program Files/graalvm...).

post instalation test:
```
Administrator@WIN MINGW64 ~/tornado/TornadoVM/bin/sdk/bin
# tornado -m tornado.benchmarks/uk.ac.manchester.tornado.benchmarks.BenchmarkRunner --params="saxpy"
WARNING: Using incubator modules: jdk.incubator.concurrent, jdk.incubator.vector
bm=saxpy-101-16777216, id=java-reference      , average=4,569409e+07, median=4,157580e+07, firstIter
ation=4,814460e+07, best=3,867590e+07
bm=saxpy-101-16777216, device=0:0  , average=2,234642e+08, median=2,204671e+08, firstIteration=6,632
591e+08, best=2,134499e+08, speedupAvg=0,2045, speedupMedian=0,1886, speedupFirstIteration=0,0726, C
V=5,1304%, deviceName= [NVIDIA CUDA] -- GeForce GT 710

Administrator@WIN MINGW64 ~/tornado/TornadoVM/bin/sdk/bin
# tornado -m tornado.benchmarks/uk.ac.manchester.tornado.benchmarks.BenchmarkRunner --params="addImage"
WARNING: Using incubator modules: jdk.incubator.vector, jdk.incubator.concurrent
bm=add-image-101-2048-2048, id=java-reference      , average=4,914668e+07, median=4,915050e+07, firs
tIteration=1,454421e+08, best=4,789100e+07
bm=add-image-101-2048-2048, device=0:0  , average=2,918137e+08, median=2,885015e+08, firstIteration=
7,450759e+08, best=2,828873e+08, speedupAvg=0,1684, speedupMedian=0,1704, speedupFirstIteration=0,19
52, CV=3,7983%, deviceName= [NVIDIA CUDA] -- GeForce GT 710

Administrator@WIN MINGW64 ~/tornado/TornadoVM/bin/sdk/bin
# tornado -m tornado.benchmarks/uk.ac.manchester.tornado.benchmarks.BenchmarkRunner --params="stencil"
WARNING: Using incubator modules: jdk.incubator.concurrent, jdk.incubator.vector
bm=stencil-101-1048576, id=java-reference      , average=6,179338e+05, median=6,121000e+05, firstIte
ration=1,267530e+07, best=5,878000e+05
bm=stencil-101-1048576, device=0:0  , average=2,885811e+06, median=2,911000e+06, firstIteration=2,92
9376e+08, best=1,865000e+06, speedupAvg=0,2141, speedupMedian=0,2103, speedupFirstIteration=0,0433,
CV=21,4355%, deviceName= [NVIDIA CUDA] -- GeForce GT 710

Administrator@WIN MINGW64 ~/tornado/TornadoVM/bin/sdk/bin
# tornado -m tornado.examples/uk.ac.manchester.tornado.examples.compute.MatrixMultiplication1D
WARNING: Using incubator modules: jdk.incubator.concurrent, jdk.incubator.vector
Computing MxM of 512x512
        Single Threaded CPU Execution: 1,11 GFlops, Total time = 242 ms
        TornadoVM Execution on GPU (Accelerated): 7,06 GFlops, Total Time = 38 ms
        Speedup: 6.368421052631579x
        Verification true


Administrator@WIN MINGW64 ~/tornado/TornadoVM/bin/sdk/bin
#

```

**Reply by lucas8338** [reply]
with few changes of course.
instead of `make graal-jdk-11-plus BACKEND=ptx,opencl` of step 6
i used
`make graal-jdk-17-plus BACKEND=opencl`

**Reply by jjfumero** [reply]
So is there any issue? As far as I understand from your comments, seems to work fine.
